### ISSUE TITLE
Pandora station in Mopidy playlist

### ISSUE THREAD

**Issue by rgov** [rectalogic/mopidy-pandora#58]
I'd like to use Mopidy-AlarmClock to play a Pandora station at a certain time of day. It only understands Mopidy playlists, so I'd need to define a playlist file that references the Pandora station to play.

**Comment by rgov** [comment]
I think I might be able to accomplish this with a `pandora:` URL. I played a station (via Moped) and saw this in the URL:

```
pandora%253Astation%253A3518796032994790657%253A3518796032994790657/Pandora%252FToday's%2520Alternative%2520Radio*
```

Decoding this twice yields:

```
pandora:station:3518796032994790657:3518796032994790657/Pandora/Today's Alternative Radio*
```

If I stick that into a .m3u file and ask Mopidy to play it, I get:

```
2017-02-25 22:50:48,212 ERROR [693:PandoraBackend-3] mopidy_pandora.playback: Error changing Pandora track: failed to lookup 'pandora:station:3518796032994790657:3518796032994790657/Pandora/Today's Alternative Radio*'.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mopidy_pandora/playback.py", line 63, in change_track
    self.change_pandora_track(track)
  File "/usr/local/lib/python2.7/dist-packages/mopidy_pandora/playback.py", line 42, in change_pandora_track
    pandora_track = self.backend.library.lookup_pandora_track(track.uri)
  File "/usr/local/lib/python2.7/dist-packages/mopidy_pandora/library.py", line 192, in lookup_pandora_track
    return self.pandora_track_cache[uri].track
  File "/usr/local/lib/python2.7/dist-packages/cachetools/lru.py", line 16, in __getitem__
    value = cache_getitem(self, key)
  File "/usr/local/lib/python2.7/dist-packages/cachetools/cache.py", line 44, in __getitem__
    return self.__missing__(key)
  File "/usr/local/lib/python2.7/dist-packages/cachetools/cache.py", line 71, in __missing__
    value = self.__missing(key)
  File "/usr/local/lib/python2.7/dist-packages/cachetools/cache.py", line 90, in __missing
    raise KeyError(key)
KeyError: "pandora:station:3518796032994790657:3518796032994790657/Pandora/Today's Alternative Radio*"
2017-02-25 22:50:48,226 WARNING [693:Core-8] mopidy.core.tracklist: Track is not playable: pandora:station:3518796032994790657:3518796032994790657/Pandora/Today's Alternative Radio*
```

**Comment by jcass77** [comment]
There are probably two things to consider here:

**1. On station URI's in general:**

The format for a Pandora station URI is ``pandora:station:station_id:token``

Although the Pandora API specification makes provision for both ``station_id`` and ``token``, my experience is that they are generally identical. However there is no guarantee that this will continue to be the case, so we need to provide for both even if that makes the URI a little more verbose than it probably needs to be.

[Mopidy-MusicBox-Webclient](https://github.com/pimusicbox/mopidy-musicbox-webclient) also has a feature that allows you to get a track's detailed information, including its URI, by activating the track's context menu. A ``track`` URI can always be changed into a ``station`` URI by substituting the 'track' scheme for 'station', and duplicating the track's ``station_id`` field into the ``token`` field.

For example, the corresponding station URI for track ``pandora:track:123:4567890`` is ``pandora:station:123:123``.

**2. On adding station URIs to Mopidy playlists and the Mopidy tracklist:**

Mopidy expects all URI's that appear in a playlist or the tracklist to point to either a stream or an actual playable track. A Pandora station URI, by comparison, represents neither a stream nor a track that can be played directly, which is why Mopidy-Pandora does not provide this feature at present.

Adding Pandora tracks to a playlist does not make sense either since track URI's expire after a short while, and will most likely not be available anymore by the time that users try to access the playlist again.

It might be possible to offer some kind of workaround for all of this by just retrieving and playing the relevant tracks when a station URI is encountered (like you tried to do in #59). Doing so will however mean departing from the standard Mopidy behavior even further, and it is difficult to predict all of the weird and wonderful ways in which something like this might be incompatible with the event sequences that Mopidy and the various frontend UI's expect.

**Comment by jcass77** [comment]
Closed via f29d1dc3637b3fc6eb62a5678cec689ade31b31f.
